agent/engine/dockerstate/docker_task_engine_state.go

```diff
@@ -457,3 +481,70 @@ func (state *DockerTaskEngineState) GetTaskByIPAddress(addr string) (string, boo
 	taskARN, ok := state.ipToTask[addr]
 	return taskARN, ok
 }
+
+// GenerateAndAddV3EndpointID generates a container identifier to be used in v3 metadata endpoint,
```

Comment:
I don't really think this is appropriately factored.  This function is doing several pretty distinct things, mixing abstraction layers, and is not really in the right place inside `DockerTaskEngineState`.  Here are the three things it does:

1. Generate an ID
2. Modify the container configuration
3. Update the mapping from ID to task and ID to container

I think 3 is the only part that's really appropriate to be inside `DockerTaskEngineState`.  Just looking at prior examples of us doing things like this, I think the most appropriate place (right now) to modify the container configuration (2 in my list above) is in `PostUnmarshalTask`; that's where we do things like injecting the credential endpoint.  1 can also likely be done there, and then you can read the `container.V3EndpointID` in this function (instead of writing to it).

Reply:
i put them together because if i generate an ID for a container but not add it to state, there's nothing to prevent another container generating a duplicate ID. i can separate the function into two, but it doesn't seem that i can separate them in two packages

Reply:
In general, we can assume that the probability of a UUID being duplicated is negligible: https://en.wikipedia.org/wiki/Universally_unique_identifier.

Reply:
I agree that we can just ignore the case where UUID is duplicate, according the [doc](https://github.com/aws/amazon-ecs-agent/blob/c0defea9598061451cf5ffd452d2ffe3d91dbc5e/agent/vendor/github.com/pborman/uuid/version4.go#L16), the chance that we get a duplicate is already small enough, considering the uniqueness is only required within each Agent - that makes the duplicate almost impossible.

Agent has already use UUID without detecting the duplicate.